### PR TITLE
Add GenServer.async/2

### LINKED ISSUE #12436: Add GenServer.async/2
Erlang/OTP 23.x added the ability to send request and receive responses from a GenServer. This is feature is basically breaking a `GenServer.call/2` apart.

In Elixir, we can implement this functionality using tasks. However, we will need to add a new field to the Task, called `tag`, to match on the reply and still keep the `ref` for the DOWN message. This is a concern because someone may have written this code:

```elixir
    def handle_info({ref, reply}, state) when state.task.ref == ref do
      ...

    def handle_info({:DOWN, ref, _, _, _}, state) when state.task.ref == ref do
      ...
```

However, with tags, the first clause will have to be rewritten to:

```elixir
    def handle_info({tag, reply}, state) when state.task.tag == tag do
      ...
```

While this looks like a breaking change, it isn't in practice because for all existing tasks APIs, `ref == tag`, but that won't be true for tasks returned by `GenServer.async/2`, so you will need to update once you incorporate the new tasks.

The other concern is related to the API. `async_call/2` is confusing because we say "calls are synchronous". We probably want to call it `task_call/2`, `async_request/2`, `request/2` or similar. Suggestions are welcome! Edit: renamed to `async/2` per discussion below.

### PR DESCRIPTION
**Pull request by whatyouhide**
Closes #12436.

### CODE REVIEW

File: lib/elixir/lib/task.ex [hunk 1]
```diff
@@ -813,17 +832,18 @@ defmodule Task do
 
   """
   @spec await(t, timeout) :: term
-  def await(%Task{ref: ref, owner: owner} = task, timeout \\ 5000) when is_timeout(timeout) do
+  def await(%Task{ref: ref, owner: owner} = task, timeout \\ 5000)
```
Comment (josevalim) [inline lib/elixir/lib/task.ex:835]: We will have to update the example given in this function. :)

**Review by josevalim — COMMENTED** [review]
Beautiful, one comment and ship it!

File: lib/elixir/lib/task.ex (hunk 1 above)
Comment (whatyouhide) [reply lib/elixir/lib/task.ex:843]: Done, I think the implementation I put in the example is ok? Any suggestions?

Comment (whatyouhide) [reply lib/elixir/lib/task.ex:843]: Done, I think the implementation I put in the example is ok? Any suggestions?

Comment (whatyouhide) [reply lib/elixir/lib/task.ex:843]: Done, I think the implementation I put in the example is ok? Any suggestions?

File: lib/elixir/lib/task.ex [hunk 2]
```diff
@@ -768,35 +786,44 @@ defmodule Task do
...
         end
 
         # If the task succeeds...
-        def handle_info({ref, result}, state) do
-          # The task succeed so we can demonitor its reference
+        def handle_info({tag, result}, state) do
+          ref = Map.fetch!(state.task_tags, tag)
+
+          # The task succeeded so we can demonitor its reference.
           Process.demonitor(ref, [:flush])
```
Comment (whatyouhide) [inline lib/elixir/lib/task.ex:811]: I think we don't need this, given `Task` itself demonitors (with `alias: reply_demonitor`)?

Comment (josevalim) [reply lib/elixir/lib/task.ex:811]: Good call.

File: lib/elixir/lib/gen_server.ex [hunk 3]
```diff
@@ -1236,6 +1241,53 @@ defmodule GenServer do
...
+
+  """
+  @doc since: "1.15.0"
+  @spec async(server, term) :: Task.t()
+  def async(server, request) do
+    case whereis(server) do
+      nil ->
+        exit({:noproc, {__MODULE__, :async, [server, request]}})
+
+      # TODO: remove this clause when we require Erlang/OTP 25+
```
Comment (josevalim) [inline lib/elixir/lib/gen_server.ex:1279]:
```suggestion
```

We are bypassing the GenServer, so we need to keep it :)

File: lib/elixir/lib/task.ex [hunk 4]
```diff
@@ -768,35 +786,44 @@ defmodule Task do
...
-          url = ...
+          url = "..."
           task = Task.Supervisor.async_nolink(MyApp.TaskSupervisor, fn -> fetch_url(url) end)
 
-          # After we start the task, we store its reference and the url it is fetching
-          state = put_in(state.tasks[task.ref], url)
+          # After we start the task, we store its reference and the URL it is fetching
+          state = put_in(state.task_refs[task.ref], url)
+          # We also store its tag and map it to the reference
+          state = put_in(state.task_tags[task.tag], task.ref)
```
Comment (josevalim) [inline lib/elixir/lib/task.ex:801]:
```suggestion
          state = put_in(state.task_refs[task.ref], {url, task.tag})
          # We also store its tag and map it to the reference
          state = put_in(state.task_tags[task.tag], {url, task.ref})
```

Or we store the url in only one of them but it is important for tag to point to ref and vice-versa.

File: lib/elixir/lib/task.ex [hunk 5]
```diff
@@ -768,35 +786,44 @@ defmodule Task do
...
 
           {:reply, :ok, state}
         end
 
         # If the task succeeds...
-        def handle_info({ref, result}, state) do
-          # The task succeed so we can demonitor its reference
+        def handle_info({tag, result}, state) do
```
Comment (josevalim) [inline lib/elixir/lib/task.ex:807]:
```suggestion
        def handle_info({tag, result}, state) when is_map_key(state.task_tags, tag) do
```

File: lib/elixir/lib/task.ex [hunk 6]
```diff
@@ -768,35 +786,44 @@ defmodule Task do
...
         end
 
         # If the task fails...
         def handle_info({:DOWN, ref, _, _, reason}, state) do
           {url, state} = pop_in(state.tasks[ref])
-          IO.puts "URL #{inspect url} failed with reason #{inspect(reason)}"
+
+          # We also clean up the tag for this task.
+          state =
+            update_in(state.task_tags, fn tags -> Map.filter(tags, &match?({_tag, ^ref}, &1)) end)
```
Comment (josevalim) [inline lib/elixir/lib/task.ex:824]: Let's keep cross references to avoid map traversals.

**Pull request merged**